cmd/zfs/zfs_project.c

```diff
@@ -0,0 +1,677 @@
+				zfs_project_usage();
+			}
+
+			dironly = B_TRUE;
+			break;
+		case 'k':
+			keep = B_TRUE;
+			break;
+		case 'R':
+			if (dironly) {
+				(void) fprintf(stderr, gettext("NOT allow "
+				    "to specify '-d' and '-R' together\n"));
+				zfs_project_usage();
+			}
+
+			recursive = B_TRUE;
+			break;
+		case 'p':
+			projid = atoll(optarg);
+			if (projid == ZFS_INVALID_PROJID) {
+				(void) fprintf(stderr,
+				    gettext("invalid project ID\n"));
+				zfs_project_usage();
+			}
+
+			break;
+		case '0':
+			newline = B_FALSE;
+			break;
+		default:
+			(void) fprintf(stderr, gettext("invalid option '%c'\n"),
+			    optopt);
+			zfs_project_usage();
+		}
+	}
+
+	switch (op) {
+	case ZFS_PROJECT_CHECK:
+		if (!recursive) {
+			if (projid != ZFS_INVALID_PROJID) {
+				(void) fprintf(stderr, gettext(
+				    "'-p' need to be together with '-c -R'\n"));
+				zfs_project_usage();
+			}
+			if (!newline) {
+				(void) fprintf(stderr, gettext(
+				    "'-0' need to be together with '-c -R'\n"));
+				zfs_project_usage();
+			}
+		}
+		if (keep) {
+			(void) fprintf(stderr,
+			    gettext("'-k' is useless together with '-c'\n"));
+			zfs_project_usage();
+		}
+		break;
+	case ZFS_PROJECT_CLEAR:
+		if (!newline) {
+			(void) fprintf(stderr,
+			    gettext("'-0' is useless together with '-C'\n"));
+			zfs_project_usage();
+		}
+		if (dironly) {
+			(void) fprintf(stderr,
+			    gettext("'-d' is useless together with '-C'\n"));
+			zfs_project_usage();
+		}
+		if (projid != ZFS_INVALID_PROJID) {
+			(void) fprintf(stderr,
+			    gettext("'-p' is useless together with '-C'\n"));
+			zfs_project_usage();
+		}
+		break;
+	case ZFS_PROJECT_SET:
+		if (!newline) {
+			(void) fprintf(stderr,
+			    gettext("'-0' is useless together with '-s'\n"));
+			zfs_project_usage();
+		}
+		if (dironly) {
+			(void) fprintf(stderr,
+			    gettext("'-d' is useless together with '-s'\n"));
+			zfs_project_usage();
+		}
+		if (keep) {
+			(void) fprintf(stderr,
+			    gettext("'-k' is useless together with '-s'\n"));
+			zfs_project_usage();
+		}
+		break;
+	default:
+		if (projid == ZFS_INVALID_PROJID) {
+			(void) fprintf(stderr,
+			    gettext("NOT specify operation type\n"));
+			zfs_project_usage();
+		}
+		if (!newline) {
+			(void) fprintf(stderr,
+			    gettext("'-0' is useless together with '-p'\n"));
+			zfs_project_usage();
+		}
+		if (dironly) {
+			(void) fprintf(stderr,
+			    gettext("'-d' is useless together with '-p'\n"));
+			zfs_project_usage();
+		}
+		if (keep) {
+			(void) fprintf(stderr,
+			    gettext("'-k' is useless together with '-p'\n"));
+			zfs_project_usage();
+		}
+		if (recursive) {
+			(void) fprintf(stderr,
+			    gettext("please add '-s' for set tree quota\n"));
+			zfs_project_usage();
+		}
+		break;
+	}
+
+	argv += optind;
+	argc -= optind;
+	if (argc == 0) {
+		(void) fprintf(stderr, gettext("miss operate target(s)\n"));
+		zfs_project_usage();
+	}
+
+	for (i = 0; i < argc && ret == 0; i++) {
```

Comment:
It is not impossible that the "zfs project" works on dataset. But I am thinking whether it is a reasonable use case for the whole FS within single tree quota. If that is true, then it is equal to "zfs project $mount_point". Since we can handle the latter case, then whether it is necessary to handle dataset parameters or not?

One of the benefit of dataset parameter for "zfs project" is that: when we support /etc/projects in the future, we may setup/check/clear multiple tree quota according to the /etc/projects via dataset parameter. So I think we can consider to support that together with /etc/projects in the future.

How do you think?